### ISSUE TITLE
Generalize `Synapses` to general events

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2#488]
Without much effort we can adapt `Synapses` so that it works with the general event types implemented for #105. This flexibility might be useful and I don't see any reason not to allow it. A question is what the syntax should look like (nothing should change for the standard use case, of course).
Here's a proposal for a syntax using a dictionary that would also work for several pathways:
```Python
G = NeuronGroup(N, '...', events={'myevent': (condition, statements)})
S = Synapses(G, G, '...', pre='...', events={'pre': 'myevent'})
```

**Comment by thesamovar** [comment]
What about each Synapses object only applies to one event type, and there is an ``event_type='spikes'`` keyword? Or does this miss some generality?

**Comment by mstimberg** [comment]
Ah, I see. It would make it slightly less general, I think. I don't have a clear use case in mind but imagine something like this:
```Python
G = NeuronGroup(N, '...', threshold='v>v_th',
                events={'trigger': ('x > x_th', 'x = 0')})
S = Synapses(G, G, 'w : 1', pre={'transmit': 'v += w', 'modulate': 'w *= modulation'},
             events={'modulate': 'trigger'})
```
Since both pathways need access to the synaptic weight `w`, it would be the easiest to have them in the same `Synapses` object.

**Comment by thesamovar** [comment]
Ah yes, OK, that syntax works for me then. We could have ``events='trigger'`` for the special case where all pathways use the same event type, and then we get the best of both worlds.

**Comment by mstimberg** [comment]
Closed via #487
